### ISSUE TITLE
The network will not converge when backend is 'cupy'

### ISSUE THREAD

**Issue by fangwei123456** [fangwei123456/spikingjelly#106]
https://github.com/fangwei123456/spikingjelly/blob/8fded1a8f1d4fa9987387ddaf6b7fb47b38941cc/spikingjelly/clock_driven/examples/speechcommands.py#L303

**Comment by fangwei123456** [comment]
```
            out_spikes_counter_frequency = net(audios)
            out1 = out_spikes_counter_frequency.data.clone()

            loss = criterion(out_spikes_counter_frequency, labels)
            loss.backward()
            gd1 = net.conv[0].weight.grad.clone()

            reset_net(net)
            for m in net.modules():
                if isinstance(m, neuron.MultiStepLIFNode):
                    m.backend = 'cupy'

            optimizer.zero_grad()
            out_spikes_counter_frequency = net(audios)
            out2 = out_spikes_counter_frequency.data.clone()
            loss = criterion(out_spikes_counter_frequency, labels)
            loss.backward()
            gd2 = net.conv[0].weight.grad.clone()
            print((gd1 - gd2).abs().max())
```

The gradients calculated by cupy and torch are almost the same:
```
Epoch 0
tensor(6.2615e-09, device='cuda:0')
tensor(5.8637e-09, device='cuda:0')
tensor(4.7331e-09, device='cuda:0')
tensor(5.8009e-09, device='cuda:0')
tensor(5.2828e-09, device='cuda:0')
tensor(3.3466e-09, device='cuda:0')
tensor(2.9905e-09, device='cuda:0')
tensor(6.0736e-09, device='cuda:0')
tensor(6.2346e-09, device='cuda:0')
```

**Comment by fangwei123456** [comment]
```
        for audios, labels in tqdm(train_dataloader, disable=True):
            audios = audios.cuda(non_blocking=True)
            labels = labels.cuda(non_blocking=True)
            # optimizer.zero_grad()

            for m in net.modules():
                if isinstance(m, neuron.MultiStepLIFNode):
                    m.backend = 'cupy'


            out_spikes_counter_frequency = net(audios)

            loss = criterion(out_spikes_counter_frequency, labels)

            loss.backward()

            gd_cupy = []
            for m in net.parameters():
                gd_cupy.append(m.grad.clone())

            reset_net(net)

            for m in net.modules():
                if isinstance(m, neuron.MultiStepLIFNode):
                    m.backend = 'torch'

            out_spikes_counter_frequency = net(audios)

            loss = criterion(out_spikes_counter_frequency, labels)

            loss.backward()

            gd_torch = []
            for m in net.parameters():
                gd_torch.append(m.grad.clone())

            reset_net(net)

            for i in range(gd_cupy.__len__()):
                print(i, gd_cupy[i].shape, (gd_cupy[i] - gd_torch[i]).abs().max().item())



            # Rate-based output decoding
            correct_rate = (out_spikes_counter_frequency.argmax(
                dim=1) == labels).float().mean().item()
            print(net.train_times, 'acc=', correct_rate)

            net.train_times += 1

```
```

Epoch 0
0 torch.Size([64, 1, 4, 3]) 2.8523226447418892e-08
1 torch.Size([64, 64, 4, 3]) 1.7923906625583186e-07
2 torch.Size([64, 64, 4, 3]) 1.9376933391157536e-09
3 torch.Size([12, 2560]) 0.0
4 torch.Size([12]) 0.10307404398918152
0 acc= 0.0625
0 torch.Size([64, 1, 4, 3]) 1.4616055921123916e-08
1 torch.Size([64, 64, 4, 3]) 1.0162838748328795e-07
2 torch.Size([64, 64, 4, 3]) 3.7057681367969053e-09
3 torch.Size([12, 2560]) 0.0
4 torch.Size([12]) 0.08879263699054718
1 acc= 0.0625
0 torch.Size([64, 1, 4, 3]) 1.9362811798373514e-08
1 torch.Size([64, 64, 4, 3]) 1.197510641759436e-07
2 torch.Size([64, 64, 4, 3]) 3.1186346838296686e-09
3 torch.Size([12, 2560]) 0.0
4 torch.Size([12]) 0.08681721985340118
2 acc= 0.171875
0 torch.Size([64, 1, 4, 3]) 2.0726254490455176e-08
1 torch.Size([64, 64, 4, 3]) 1.1302182656436344e-07
2 torch.Size([64, 64, 4, 3]) 3.0724209842958317e-09
3 torch.Size([12, 2560]) 0.0
4 torch.Size([12]) 0.05158136785030365
3 acc= 0.0625
0 torch.Size([64, 1, 4, 3]) 1.3601729520473782e-08
1 torch.Size([64, 64, 4, 3]) 3.859836894548607e-08
2 torch.Size([64, 64, 4, 3]) 2.7364497334758653e-09
3 torch.Size([12, 2560]) 0.0
4 torch.Size([12]) 0.07391665875911713
4 acc= 0.046875
0 torch.Size([64, 1, 4, 3]) 1.6029069627165882e-08
1 torch.Size([64, 64, 4, 3]) 1.0936236094494234e-07
2 torch.Size([64, 64, 4, 3]) 2.981346725050571e-09
3 torch.Size([12, 2560]) 0.0
4 torch.Size([12]) 0.057681918144226074
5 acc= 0.03125
0 torch.Size([64, 1, 4, 3]) 1.4526392533298349e-08
1 torch.Size([64, 64, 4, 3]) 9.547437684886972e-08
2 torch.Size([64, 64, 4, 3]) 1.9769352821441544e-09
3 torch.Size([12, 2560]) 0.0
4 torch.Size([12]) 0.05814585089683533
6 acc= 0.0625
```

**Comment by fangwei123456** [comment]
```
import torch

device = 'cuda:0'
x = torch.rand([2, 3, 4], device=device).transpose(0, 2)
y = torch.zeros_like(x)
args_list = [x, y]
ret_list = []
for item in args_list:
    if isinstance(item, torch.Tensor):
        item = item.contiguous()
        ret_list.append(item.data_ptr())
print(ret_list)
```

`[47284487168, 47284487168]`

**Comment by fangwei123456** [comment]
https://discuss.pytorch.org/t/bug-two-tensors-have-the-same-data-ptr/132104
